crates/oxc_transformer/src/proposals/decorators.rs

```diff
@@ -0,0 +1,149 @@
+use std::rc::Rc;
+
+use oxc_allocator::Box;
+use oxc_ast::{ast::*, AstBuilder};
+use oxc_span::{Atom, SPAN};
+use oxc_syntax::operator::{AssignmentOperator, LogicalOperator};
+use serde::Deserialize;
+
+use crate::{context::TransformerCtx, options::TransformOptions};
+
+/// Proposal: Decorators
+///
+/// References:
+/// * <https://babel.dev/docs/babel-plugin-proposal-decorators>
+/// * <https://github.com/babel/babel/blob/main/packages/babel-plugin-proposal-decorators>
+/// * <https://github.com/tc39/proposal-decorators>
+pub struct Decorators<'a> {
+    ast: Rc<AstBuilder<'a>>,
+    _ctx: TransformerCtx<'a>,
+    options: DecoratorsOptions,
+    // Insert to the top of the program
+    top_statements: Vec<Statement<'a>>,
+}
+
+#[derive(Debug, Clone, Copy, Default, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DecoratorsOptions {
+    version: Version,
+}
+
+#[derive(Debug, Default, Clone, Copy, Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum Version {
+    Legacy,
+    #[serde(rename = "2023-05")]
+    #[default]
+    Year2023May,
+}
+impl Version {
+    fn is_legacy(self) -> bool {
+        matches!(self, Self::Legacy)
+    }
+}
+
+impl<'a> Decorators<'a> {
+    pub fn new(
+        ast: Rc<AstBuilder<'a>>,
+        ctx: TransformerCtx<'a>,
+        options: &TransformOptions,
+    ) -> Option<Self> {
+        options.decorators.map(|options| Self {
+            ast,
+            _ctx: ctx,
+            options,
+            top_statements: Vec::new(),
+        })
+    }
+
+    pub fn transform_program(&mut self, program: &mut Program<'a>) {
+        program.body.splice(0..0, self.top_statements.drain(..));
```

Comment:
Should we consider directives like "use strict"?

Reply:
directives are in another field for the exact situations like this ;-)

https://github.com/oxc-project/oxc/blob/fda2d7c334b0ec5dd57656fd8ae1601460594af2/crates/oxc_ast/src/ast/js.rs#L24

Reply:
I don't get your questions.

`super` is a special expression 🤔 

https://github.com/oxc-project/oxc/blob/fda2d7c334b0ec5dd57656fd8ae1601460594af2/crates/oxc_ast/src/ast/js.rs#L56